File: library/android/src/androidHostTest/kotlin/ApplyDynamicPriceTests.kt
package com.adsbynimbus.dynamicprice

import android.os.SystemClock
import com.adsbynimbus.Nimbus
import com.adsbynimbus.NimbusResponse.Bid.MarkupType
import com.google.android.libraries.ads.mobile.sdk.common.AdRequest
import io.mockk.*
import kotlin.test.*

class ApplyDynamicPriceTests {
    @BeforeTest
    fun setUp() {
        mockkStatic(SystemClock::class)
        every { SystemClock.elapsedRealtime() } returns 0
    }

    @AfterTest
    fun tearDown() {
        unmockkStatic(SystemClock::class)
    }

    @Test fun `applyDynamicPrice sets customTargeting for static ads`() {
        val ad = createNimbusAd(type = MarkupType.Banner)
        val request = AdRequest.Builder("id").apply {
            ad.applyDynamicPrice(this, mapping = mapping)
        }.build()

        assertEquals(ad.id, actual = request.customTargeting["na_id"])
        assertEquals(mapping.getTarget(ad), actual = request.customTargeting["na_bid"])
        assertEquals(ad.bid.ext.omp?.buyer, actual = request.customTargeting["na_network"])
        assertEquals("${ad.bid.w}x${ad.bid.h}", actual = request.customTargeting["na_size"])
        assertEquals("static", actual = request.customTargeting["na_type"])
        assertEquals("static", actual = request.customTargeting["na_render"])
        assertNull(request.customTargeting["na_bid_video"])
    }

    @Test fun `applyDynamicPrice sets customTargeting for video ads`() {
        val ad = createNimbusAd(type = MarkupType.Video)
        val request = AdRequest.Builder("id").apply {
            ad.applyDynamicPrice(this, mapping = mapping)
        }.build()

        assertEquals(ad.id, actual = request.customTargeting["na_id"])
        assertEquals(mapping.getTarget(ad), actual = request.customTargeting["na_bid_video"])
        assertEquals(ad.bid.ext.omp?.buyer, actual = request.customTargeting["na_network"])
        assertEquals("0x0", actual = request.customTargeting["na_size"])
        assertEquals("video", actual = request.customTargeting["na_type"])
        assertEquals("video", actual = request.customTargeting["na_render"])
        assertNull(request.customTargeting["na_bid"])
    }

    @Test fun `applyDynamicPrice does not modify other customTargeting values`() {
        val ad = createNimbusAd()
        val request = AdRequest.Builder("id").apply {
            putCustomTargeting("test_key", "test_value")
            ad.applyDynamicPrice(this, mapping = mapping)
        }.build()

        assertContains(request.customTargeting, key = "test_key")
        assertEquals("test_value", actual = request.customTargeting["test_key"])
        assertEquals("auctionId-0", actual = request.customTargeting["na_id"])
        assertEquals("200", actual = request.customTargeting["na_bid"])
        assertEquals("network", actual = request.customTargeting["na_network"])
        assertEquals("320x50", actual = request.customTargeting["na_size"])
        assertEquals("static", actual = request.customTargeting["na_type"])
        assertEquals("static", actual = request.customTargeting["na_render"])
        assertNull(request.customTargeting["na_bid_video"])
    }

    @Test fun `applyDynamicPrice sets na_bid=0 when Nimbus testMode is true`() {
        Nimbus.configuration.testMode = true
        val ad = createNimbusAd(type = MarkupType.Banner)
        val request = AdRequest.Builder("id").apply {
            ad.applyDynamicPrice(this, mapping = mapping)
        }.build()

        assertEquals("0", actual = request.customTargeting["na_bid"])
        assertNull(request.customTargeting["na_bid_video"])

        Nimbus.configuration.testMode = false
    }

    @Test fun `applyDynamicPrice uses na_bid_video=0 when Nimbus testMode is true`() {
        Nimbus.configuration.testMode = true
        val ad = createNimbusAd(type = MarkupType.Video)
        val request = AdRequest.Builder("id").apply {
            ad.applyDynamicPrice(this, mapping = mapping)
        }.build()

        assertEquals("0", actual = request.customTargeting["na_bid_video"])
        assertNull(request.customTargeting["na_bid"])

        Nimbus.configuration.testMode = false
    }
}

File: library/ios/src/test/swift/ApplyDynamicPriceTests.swift
//
//  NimbusAd+TargetingTests.swift
//  DynamicPriceTests
//
//  Created on 8/12/22.
//  Copyright © 2026 Nimbus Advertising Solutions Inc. All rights reserved.
//

import GoogleMobileAds
import NimbusKit
import Testing

import DynamicPrice

@Suite(.serialized) struct ApplyDynamicPriceTests {

    @Test func `applyDynamicPrice sets customTargeting for static ads`() async {
        let ad = createNimbusAd(type: .static)
        let request = AdManagerRequest()
        ad.applyDynamicPrice(request, mapping: mapping)

        #expect(request.customTargeting?["na_id"] as? String == ad.id)
        #expect(request.customTargeting?["na_bid"] as? String == mapping.getTarget(ad))
        #expect(request.customTargeting?["na_network"] as? String == ad.bid.ext?.omp?.buyer)
        #expect(request.customTargeting?["na_size"] as? String == "\(ad.bid.w!)x\(ad.bid.h!)")
        #expect(request.customTargeting?["na_type"] as? String == "static")
        #expect(request.customTargeting?["na_render"] as? String == "static")
        #expect(request.customTargeting?["na_bid_video"] == nil)
    }

    @Test func `applyDynamicPrice sets customTargeting for video ads`() async {
        let ad = createNimbusAd(type: .video)
        let request = AdManagerRequest()

        ad.applyDynamicPrice(request, mapping: mapping)

        #expect(request.customTargeting?["na_id"] as? String == ad.id)
        #expect(request.customTargeting?["na_network"] as? String == ad.bid.ext?.omp?.buyer)
        #expect(request.customTargeting?["na_size"] as? String == "0x0")
        #expect(request.customTargeting?["na_bid_video"] as? String == mapping.getTarget(ad))
        #expect(request.customTargeting?["na_type"] as? String == "video")
        #expect(request.customTargeting?["na_render"] as? String == "video")
    }

    @Test func `applyDynamicPrice does not modify other customTargeting values`() async {
        let ad = createNimbusAd(type: .static)
        let request = AdManagerRequest()
        request.customTargeting = [:]
        request.customTargeting?["test_key"] = "test_value"

        ad.applyDynamicPrice(request, mapping: mapping)

        #expect(request.customTargeting?["na_id"] as? String == ad.id)
        #expect(request.customTargeting?["na_network"] as? String == ad.bid.ext?.omp?.buyer)
        #expect(request.customTargeting?["na_size"] as? String == "\(ad.bid.w!)x\(ad.bid.h!)")
        #expect(request.customTargeting?["na_bid"] as? String == "200")
        #expect(request.customTargeting?["na_type"] as? String == "static")
        #expect(request.customTargeting?["na_render"] as? String == "static")
        #expect(request.customTargeting?["test_key"] as? String == "test_value")
    }

    @MainActor
    @Test func `applyDynamicPrice uses na_bid=0 when Nimbus.shared.testMode = true`() async {
        Nimbus.configuration.testMode = true
        let ad = createNimbusAd(type: .static)
        let request = AdManagerRequest()
        request.customTargeting = [:]

        ad.applyDynamicPrice(request, mapping: mapping)

        #expect(request.customTargeting?["na_bid"] as? String == "0")
        #expect(request.customTargeting?["na_bid_video"] == nil)

        Nimbus.configuration.testMode = false
    }

    @MainActor
    @Test func `applyDynamicPrice uses na_bid_video=0 when Nimbus.shared.testMode = true`() async {
        Nimbus.configuration.testMode = true
        let ad = createNimbusAd(type: .video)
        let request = AdManagerRequest()
        request.customTargeting = [:]

        ad.applyDynamicPrice(request, mapping: mapping)

        #expect(request.customTargeting?["na_bid_video"] as? String == "0")
        #expect(request.customTargeting?["na_bid"] == nil)

        Nimbus.configuration.testMode = false
    }
}
